core/src/main/java/com/linecorp/armeria/internal/common/CancellationScheduler.java

```diff
@@ -291,213 +261,252 @@ private void setTimeoutNanosFromNow(long timeoutNanos) {
-            }
-
-            if (!whenTimedOutUpdater.compareAndSet(this, null, COMPLETED_FUTURE)) {
-                whenTimedOut.doComplete();
-            }
-        }
+    public long timeoutNanos() {
+        return isInitialized() ? timeoutNanos : pendingTimeoutNanos;
     }
 
-    public boolean isTimedOut() {
-        return state == State.TIMED_OUT;
+    public long startTimeNanos() {
+        return startTimeNanos;
     }
 
-    public long timeoutNanos() {
-        return isInitialized() ? timeoutNanos : pendingTimeoutNanos;
+    public CompletableFuture<Throwable> whenCancelling() {
+        final CancellationFuture whenCancelling = this.whenCancelling;
+        if (whenCancelling != null) {
+            return whenCancelling;
+        }
+        final CancellationFuture cancellationFuture = new CancellationFuture();
+        if (whenCancellingUpdater.compareAndSet(this, null, cancellationFuture)) {
+            return cancellationFuture;
+        } else {
+            return this.whenCancelling;
+        }
     }
 
-    private boolean isInitialized() {
-        return pendingTimeoutTask == initializedPendingTimeoutTask;
+    public CompletableFuture<Throwable> whenCancelled() {
+        final CancellationFuture whenCancelled = this.whenCancelled;
+        if (whenCancelled != null) {
+            return whenCancelled;
+        }
+        final CancellationFuture cancellationFuture = new CancellationFuture();
+        if (whenCancelledUpdater.compareAndSet(this, null, cancellationFuture)) {
+            return cancellationFuture;
+        } else {
+            return this.whenCancelled;
+        }
     }
 
+    @Deprecated
     public CompletableFuture<Void> whenTimingOut() {
         final TimeoutFuture whenTimingOut = this.whenTimingOut;
         if (whenTimingOut != null) {
             return whenTimingOut;
         }
-
         final TimeoutFuture timeoutFuture = new TimeoutFuture();
         if (whenTimingOutUpdater.compareAndSet(this, null, timeoutFuture)) {
+            whenCancelling().thenAccept(cause -> {
+                if (cause instanceof TimeoutException) {
+                    timeoutFuture.doComplete();
+                } else {
+                    timeoutFuture.doCancel();
+                }
+            });
             return timeoutFuture;
         } else {
             return this.whenTimingOut;
         }
     }
 
+    @Deprecated
     public CompletableFuture<Void> whenTimedOut() {
         final TimeoutFuture whenTimedOut = this.whenTimedOut;
         if (whenTimedOut != null) {
             return whenTimedOut;
         }
-
         final TimeoutFuture timeoutFuture = new TimeoutFuture();
         if (whenTimedOutUpdater.compareAndSet(this, null, timeoutFuture)) {
+            whenCancelled().thenAccept(cause -> {
+                if (cause instanceof TimeoutException) {
+                    timeoutFuture.doComplete();
+                } else {
+                    timeoutFuture.doCancel();
```

Comment:
IIRC we never cancel the future returned by `whenTimed/TimingOut()`, i.e. any callbacks added by a user will be invoked only when the request has timed out.